### ISSUE TITLE
abstract the swear/case/swindle contract to handle polymorphic evidence

### ISSUE THREAD

**Issue by orenyodfat** [ethersphere/swap-swear-and-swindle#3]
Viktor : 
a swear is a case state transition automata, where at each stage of the litigation, a witness contract is called with the same ABI: `testimonyFor(status, serviceId, clientAddress) -> newstatus`
and it is this contracts specific methods which you need to call to submit the evidence
this givesa a superclean and superlean swear abstraction.
Oren: 
I like the idea.
so in the basic “mirror” game there will be two 2 expert witnesses .
First one , which validate and confirm the client subscription to the service by validate the service promise ,lets call it “SubscriptionWitness”.
It introduce an abi for clients to submit a “promise” as an evident.
Second , which check the actual mirroring staff by resolving both enss and compare them ,lets call it “MirrorWitness” .
For the case of the “mirror” game the client do not need to submit any evident because it the enss addresses are hard coded in the contract.
So the swear contract will call   SubscriptionWitness.testimoryFor(status, serviceId, clientAddress) and if the status is OK it will call “MirrorWitness.testimoryFor(status, serviceId, clientAddress) ” and take decision according to that..
Further abstraction can be done by having the passing the swear contrat a list of witnesses contract addresses , so it then can iterate through each :
e.g
for (i=0;i<numberOfWitnesses;i++)
{
status = witness[i].testimoryFor(status, serviceId, clientAddress)
if status != STATUS_OK{
  return status
}
}
return status

**Comment by zelig** [comment]
Initial challanges are also statuses (valid continuations from the initial non-challanged state) 
Therefore `openCase` should be `openCase(ACCUSATION)` where `ACCUSATION` is a status.

Any litigation process then is described by a set of status transitions.

For generality I assume the `testimonyFor` transaction returns an int which represents an outcome.
Outcomes of witness testimonies are driving the transition from status to status. 
We do not need an array of witnesses, cos we can represent them with their own status.
GUILTY and NOT_GUILTY are also statuses.  
We can further stipulate that there are two end states (UNCHALLANGED and GUILTY)

openCase checks if the status is UNCHALLANGED and you can get from there to a challanged state.
and go as far as the case as possible. A testimony call can give a special outcome (PENDING = inconclusive) as return value (i recommend 0) which postpones the trial.
The trial can be restarted (after the evidence is submitted to that witness or some deadline is exceeded)

We can further stipulate that if there is no state transition found for a witness testimony, then a default contract is called as a witness.

Each challange state can have a custom grace period falling back to default deadline.

So 

DEFAULT_CONTRACT = <witnessAddess> // if not specified, this contract is used
DEFAULT_GRACE_PERIOD = <numberOfSeconds> // could also be given as number of blocks

GRACE_PERIODS // is the map from status to custom period, if not found falls back to the default DEFAULT_GRACE_PERIOD 
WITNESSES if the map from status to witness, if not found, DEFAULT_CONTRACT is called 
if the contract address is zero (not found in the WITNESSES map), this contract's `Proceed` method is called
TRANSITIONS maps from OUTCOME (expert witness testimony) and staus to new status. 
typically the possible outcomes are fewer than statuses so 

```
mapping(outcome_t => mapping(status_t => status_t) public
```

the mirror game:

```
// statuses
UNCHALLENGED   
MIRROR_CHALLENGE
VALID_MIRROR_CHALLANGE
GUILTY
NOT_GUILTY

// transitions:
UNCHALLENGED ==> MIRROR_CHALLENGE
[VALID, MIRROR_CHALLENGE] ==> VALID_MIRROR_CHALLENGE
[INVALID, MIRROR_CHALLENGE] ==> NOT_GUILTY
[VALID,  VALID_MIRROR_CHALLENGE] ==>  GUILTY
[INVALID,  VALID_MIRROR_CHALLENGE] ==>  NOT_GUILTY

// witnesses
MIRROR_CHALLENGE => <paymentValidatorContract>
VALID_MIRROR_CHALLENGE => <ENSMirrotValidatorContract>


```

Proceed on not guilty could initiate compensation to the defendant and punishment for the plaintiff

```
function getWitness() returns address {
   witness = WITNESSES[status]
   if witness == '0x0' { // fallback to default
      return DEFAULT_WITNESS
   }
   return witness
}

function getGracePeriod() returns address {
   gp = GRACE_PERIODS[status]
   if gp == '0x0' { // fallback to default
      return DEFAULT_GRACE_PERIOD
   }
   return gp 
}

function expired() {
   return time >  expiry
}

function trial() {
   if status == UNCHALLENGED {
      return
   } 
   for status != UNCHALLENGED {   
        if expired() {
             status = UNCHALLENGED
             return
        }
 
        address memory witness = getWitness();
        outcome_t memory outcome;
        if witness == 0x0 {
            outcome = proceed();
        } else {
           outcome = witness.Call(serviceId, caseId, clientAddr, plaintiffAddr)
       }
       if outcome == PENDING {
           return
      }
      status = TRANSITIONS[outcome][status]
      // continue
  }
}


```

**Comment by orenyodfat** [comment]
-"Therefore openCase should be openCase(ACCUSATION) where ACCUSATION is a status." 
Why not let the contract maintain the status(state) of each claim ? 
  Initial state for openCase for the mirror game will be MIRROR_CHALLENGE. (any other possible initial states ?)
  If then the state shift to VALID_MIRROR_CHALLENGE .(and it is pending because it needs to provide more info ..)..a client should then re openCase  after submitting  the required info to the ENSMirrotValidatorContract...the trial will continue from the state it was stopped. 

- to keep the contract general , the deploying service should submit the witnesses map to the contract before the game can begin. right?

**Comment by MrTibbles** [comment]
In berlin we discussed the possibility of a dynamic journey selection process. This could be put toward the stipulation of witnesses also. Rather than the potential of a service provider stipulating witnesses that can be coerced in some way?

**Comment by zelig** [comment]
@MrTibbles I dont understand. 
Witnesses here are deterministic contracts
A game is safe only if witnesses are explicit and immutable from the beginning

**Comment by MrTibbles** [comment]
Sorry, i wasnt clear. What i meant was that **only once on initiation** of a swear game the witnesses would be chosen from a pool of potential witnesses, much like in the real world with jury selection. So rather than the service provider stipulating who the witnesses are, they would be chosen at random, this way avoiding any favourable selection process.
A basic reputation scheme could also be created over time by keeping track of witness allocation and successful compensation cases.

**Comment by orenyodfat** [comment]
I more or less see what you are aiming to.
As I see it there are two games types: 
1. a games  were the court can take decision based on real  proven facts by the expert witnesses(which I still think it is a bit misleading term)  ,e.g the mirror game.
In this type of a game there is no place for randomly choose witnesses (it is not lottery).
2.a game were the facts can not be deterministic technology proven. e.g  ownership of an asset which is not on the blockchain or something similar..
In this type of game there might be a place to choose expert witnesses (or accept there opinion) according to their reputation.

**Comment by MrTibbles** [comment]
It would never be a lottery in regards to the mirror game, as it is a binary outcome, the disputed content has either been included or it hasnt. You say it yourself, "I have a proven evident in my hand".

**Comment by MrTibbles** [comment]
Going back to @zelig 's comment
> Witnesses here are deterministic contracts
Are these witness contracts intended to be persistent? If they are not intended to be used more than once, then there is no need to be selecting them at random! As they will be instantiated on an adhoc basis, no?

**Comment by zelig** [comment]
@MrTibbles 

>  A game is safe only if witnesses are explicit and immutable from the beginning

> Are these witness contracts intended to be persistent? If they are not intended to be used more than once, then there is no need to be selecting them at random! As they will be instantiated on an adhoc basis, no?

I dont understand what you are asking. The witness contracts are the rules of the game. Set in stone from the beginning. I cannot even interpret the notions of 'instantiation on an adhoc basis', 'using only once', 'selecting random' .

Once again witnesses are contracts associated with a stage of litigation. They are called to evaluate an evidence submitted to them, their testimony if conclusive advances the trial by putting the swindle contract in a new stage according to the transition graph.

Witness associations can be thought of as a function: `W: N -> Address`
Transition graph can be thought of as a function `T:  <N,N> -> N`
If the trial is in state `s_i`, then the next state is `s_(i+1) = T(o, s)` where `o` is the return value of 
`W(s).testimonyFor(caseId, serviceId, clientAddress)` 
Not sure formalising it helps :)

**Comment by MrTibbles** [comment]
Im not sure how much clearer i can be. 

> Set in stone from the beginning.
>> What i meant was that only once on initiation of a swear game the witnesses would be chosen from a pool of potential witnesses, much like in the real world with jury selection.

**Comment by zelig** [comment]
@orenyodfat 

> Why not let the contract maintain the status(state) of each claim ?
Initial state for openCase for the mirror game will be MIRROR_CHALLENGE. (any other possible initial states ?)

well my motivation was to let the state also encode that there is a litigation process.
If the case contract is only created when a case is opened then this is not needed indeed.
But possible accusations (legitimate initial litigation states) still need to be listed somewhere.

In fact in the spirit of #4 we could implement a generic global abstact swindle contract which knows how to conduct a trial. Each service would register a contract which handles the specifics, ie. the witnesses, and the transition graph, punishment etc.




    to keep the contract general , the deploying service should submit the witnesses map to the contract before the game can begin. right?

**Comment by zelig** [comment]
@MrTibbles 
ok I try to interpret it. In cases when a subjective decision needs to be made, you could implement a witness contract which delegates the testimony to a jury which determines the outcome by vote.
Yes this is possible, but what usecase are you thinking of?

**Comment by MrTibbles** [comment]
Perhaps on the grounds of forcible eviction from a game due to a player opening too many invalid compensation claims? Which does rely on the basis that all claims are being logged. I think this is out of scope for now though.

**Comment by orenyodfat** [comment]
@MrTibbles @zelig "Yes this is possible, but what usecase are you thinking of?"
I guess it is for cases were the facts can not be deterministic technology proven. e.g ownership of an asset which is not on the blockchain or something similar..

**Comment by orenyodfat** [comment]
@zelig if we assume that all expert witnesses are deployed to the game by the service (which offer the game) what and when will be the case for using default contract ?

**Comment by Eknir** [comment]
Could somebody provide me with a conclusion of this discussion? 
Also: to which extent are the thoughts here implemented in the current smart-contract suite; to which extent are the thoughts here still relevant?
If not implemented and still relevant I would propose to make this discussion somewhat more publicly known and attempt to implement it.
If implemented, can I close the issue?

**Comment by ralph-pichler** [comment]
This was a discussions based on a code base that was abandoned long ago and is no longer relevant (the witness interface and the trial rules have changed significantly).  Closing this.
